### PR TITLE
Support default methods

### LINKED ISSUE #1220: Recognize annotations on default methods in implemented interface
### TestNG Version

6.9.13.6
### Expected behavior

When TestNG is invoked on a class, it should invoke any concrete (default) methods annotated with `@Test` on any interface or super-interface, just as it currently does with concrete methods in abstract classes. 

This is a feature parity request for JUnit 5 (http://junit.org/junit5/docs/current/user-guide/#writing-tests-interface-default-methods)
### Actual behavior

Default method tests are not run
### Is the issue reproducible on runner?

Tested on shell, IDEA, and buck
### Test case sample

```
interface TestMixin {
@Test 
default void test1() { }
}

class TestClass implements TestMixin {
@Test
public void test2() { }
}
```

When testing TestClass, both test1 and test2 should run and pass. Currently, only test2 is invoked.

### PR DESCRIPTION
**Pull request by d1231**
Fixes #1220  .

### Did you remember to?

- [ x ] Add test case(s)
- [ x ] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/test8/java/test/defaultmethods/DefaultMethodTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.testng.ITestNGListener;
+import org.testng.ITestNGMethod;
+import org.testng.TestNG;
+import org.testng.annotations.Test;
+import test.SimpleBaseTest;
+import test.listeners.TestAndClassListener;
+
+public class DefaultMethodTest extends SimpleBaseTest {
+
+    @Test(description = "Test default methods defined in a interface should be run")
```
Comment (cbeust) [inline src/test8/java/test/defaultmethods/DefaultMethodTest.java]: "an interface"

File: build.gradle [hunk 2]
```diff
@@ -21,7 +21,7 @@ buildscript {
     dependencies {
         classpath "io.codearte.gradle.nexus:gradle-nexus-staging-plugin:0.5.3"
         classpath 'org.hibernate.build.gradle:version-injection-plugin:1.0.0'
-	classpath 'org.sonarsource.scanner.gradle:sonarqube-gradle-plugin:2.2.1'
+        classpath 'org.sonarsource.scanner.gradle:sonarqube-gradle-plugin:2.2.1'
```
Comment (cbeust) [inline build.gradle]: Would you mind taking out all these indentation changes in the build file?

File: src/test8/java/test/defaultmethods/DefaultMethodTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    public void testDefaultShouldRun() {
+        TestNG tng = create(TestA.class);
+        final TestCounter listener = new TestCounter();
```
Comment (cbeust) [inline src/test8/java/test/defaultmethods/DefaultMethodTest.java]: Please remove `final` (and those below).

Also, it looks like all these new test methods share a lot of common code, any way you can refactor?

File: src/test8/java/test/defaultmethods/DefaultMethodTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @Test(description = "Before class default methods defined in a interface should be run")
+    public void beforeClassDefaultShouldRun() {
+        TestNG tng = create(TestA.class);
+        final TestCounter listener = new TestCounter();
+        tng.addListener((ITestNGListener) listener);
+        tng.run();
+
+        final ITestNGMethod[] beforeClassMethods = listener.testClass.getBeforeClassMethods();
```
Comment (cbeust) [inline src/test8/java/test/defaultmethods/DefaultMethodTest.java]: `final`.

**Review by cbeust — COMMENTED** [review]
Looks good with just a few minor comments.

**Comment by juherr** [comment]
If possible, I prefer 'src/test/java8' as folder name.

And currently the Java 8 tests are always run and the build fails on Java 7 (check Travis output).

**Comment by d1231** [comment]
I will fix all those issues you mention.

What should be the default behaviour regarding running the java 8 tests?

**Comment by d1231** [comment]
I fixed all the issues you mentioned.

Managed also to correct the folder layout and fix running the test when compiled with java 7. Now the test will run only if the java version is higher than or is java 8.

**Review by juherr — APPROVED** [review]
LGTM 👍 Ping @cbeust

**Pull request merged**